### ISSUE TITLE
Incorrect "Best" returned at the end of search

### ISSUE THREAD

**Issue by NickBuchny** [hyperopt/hyperopt-sklearn#68]
When performing a search instead of returning the best trial it's returning... well I'm not sure. 

Code below 

`
from hpsklearn import HyperoptEstimator
from hyperopt import fmin, tpe, hp, STATUS_OK, Trials
from sklearn.model_selection import cross_val_score

max_leaf_n = [-11, -10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

def hyperopt_train_test(params):
    clf = GradientBoostingClassifier(**params)
    return cross_val_score(clf, X_train, y_train, cv=4, scoring='roc_auc').mean()

space4gbc = {
    'learning_rate': hp.choice('learning_rate', np.array(range(1, 10))/100.), 
    'n_estimators': hp.choice('n_estimators', np.array(range(100, 10000, 100))), 
    'subsample': hp.choice('subsample', np.array(range(5, 10))/10.),
    'max_depth': hp.choice('max_depth', np.array(range(3, 11))),
    'max_features': hp.choice('max_features', np.array(range(5, 10))/10.),
    'max_leaf_nodes':hp.choice('max_leaf_nodes', np.array(max_leaf_n))

}

best = 0
def f(params):
    global best
    acc = hyperopt_train_test(params)
    if acc > best:
        best = acc
        print 'new best:', best, params
    return {'loss': -acc, 'status': STATUS_OK}

trials = Trials()
best = fmin(f, space4gbc, algo=tpe.suggest, max_evals=1500, trials=trials)
print 'best:'
print best
`

Returns:

WARN: OMP_NUM_THREADS=None =>
... If you are using openblas if you are using openblas set OMP_NUM_THREADS=1 or risk subprocess calls hanging indefinitely
new best: 0.935256874069 {'n_estimators': 6000, 'subsample': 0.5, 'max_features': 0.90000000000000002, 'max_leaf_nodes': 2, 'learning_rate': 0.089999999999999997, 'max_depth': 3}
new best: 0.95180737041 {'n_estimators': 6200, 'subsample': 0.59999999999999998, 'max_features': 0.5, 'max_leaf_nodes': -8, 'learning_rate': 0.050000000000000003, 'max_depth': 5}
new best: 0.952759609985 {'n_estimators': 1800, 'subsample': 0.90000000000000002, 'max_features': 0.69999999999999996, 'max_leaf_nodes': -10, 'learning_rate': 0.070000000000000007, 'max_depth': 8}
new best: 0.953281116353 {'n_estimators': 4000, 'subsample': 0.59999999999999998, 'max_features': 0.5, 'max_leaf_nodes': -10, 'learning_rate': 0.089999999999999997, 'max_depth': 8}
new best: 0.95399352664 {'n_estimators': 5800, 'subsample': 0.90000000000000002, 'max_features': 0.5, 'max_leaf_nodes': -4, 'learning_rate': 0.089999999999999997, 'max_depth': 9}
new best: 0.954052296118 {'n_estimators': 2000, 'subsample': 0.80000000000000004, 'max_features': 0.59999999999999998, 'max_leaf_nodes': -11, 'learning_rate': 0.040000000000000001, 'max_depth': 10}
new best: 0.954097993402 {'n_estimators': 2000, 'subsample': 0.80000000000000004, 'max_features': 0.59999999999999998, 'max_leaf_nodes': -2, 'learning_rate': 0.040000000000000001, 'max_depth': 7}
new best: 0.954157499429 {'n_estimators': 5400, 'subsample': 0.80000000000000004, 'max_features': 0.5, 'max_leaf_nodes': -11, 'learning_rate': 0.02, 'max_depth': 10}
new best: 0.954237104722 {'n_estimators': 6500, 'subsample': 0.80000000000000004, 'max_features': 0.5, 'max_leaf_nodes': -4, 'learning_rate': 0.02, 'max_depth': 10}
new best: 0.954264561746 {'n_estimators': 8400, 'subsample': 0.80000000000000004, 'max_features': 0.5, 'max_leaf_nodes': -3, 'learning_rate': 0.029999999999999999, 'max_depth': 10}
new best: 0.954437874474 {'n_estimators': 6300, 'subsample': 0.80000000000000004, 'max_features': 0.5, 'max_leaf_nodes': -3, 'learning_rate': 0.029999999999999999, 'max_depth': 7}
new best: 0.954742832472 {'n_estimators': 7100, 'subsample': 0.80000000000000004, 'max_features': 0.5, 'max_leaf_nodes': -6, 'learning_rate': 0.029999999999999999, 'max_depth': 7}
best:
{'max_leaf_nodes': 5, 'learning_rate': 2, 'n_estimators': 70, 'subsample': 3, 'max_features': 0, 'max_depth': 4}

**Comment by bjkomer** [comment]
I think what's going on here is that for choice parameters, it is outputting the index within the array that was chosen. So when it displays `'max_leaf_nodes': 5` that corresponds to the value `-6` which is in the 5th position of your array. I believe it is set up this way because `hyperopt` supports nesting of search spaces using `hp.choice` so printing the value as opposed to the index can get weird in a lot of cases.

Another point of note, you currently aren't using `hyperopt-sklearn` in your code, but pure `hyperopt` instead. There's nothing wrong with this, but if you want to try your code within the `hyperopt-sklearn` framework, it will look something more like this:

```
from hpsklearn import HyperoptEstimator, gradient_boosting

clf = gradient_boosting('my_clf',
        learning_rate=space4gbc['learning_rate'],
        n_estimators=space4gbc['n_estimators'],
        #etc...
)
estim = HyperoptEstimator(classifier=clf, algo=tpe.suggest, max_evals=1500)
estim.fit(X_train, y_train)

print(estim.trials.results)
print(estim.score(X_test, y_test))
print(estim.best_model())
```

Also, the `GradientBoostingClassifier` treats all negative values of `max_leaf_nodes` the same as `None` so the search problem will be easier if only one negative value is used.

**Comment by NickBuchny** [comment]
Ah I understand! Thanks for clearing that up. 

Yes I'm using the package directly as I'm using XGBoost also which is not listed on the on the documentation (I'm not sure if it matters or not though?).

**Comment by BenTeoNeb** [comment]
Thanks  for clearing that up. Is there an helper function to retrieve directly the values from the parameter space ?
